lib/set.cc

```diff
@@ -0,0 +1,784 @@
+		}
+		else
+		{
+			return Nan::ThrowTypeError("Expected a string, Buffer, RegExp, or RE2 instance in the pattern list.");
+		}
+
+		if (translateRegExp(data, size, flags.multiline, buffer))
+		{
+			data = &buffer[0];
+			size = buffer.size() - 1;
+		}
+
+		std::string error;
+		if (set->set.Add(re2::StringPiece(data, size), &error) < 0)
+		{
+			if (error.empty())
+			{
+				error = "Invalid pattern in RE2.Set.";
+			}
+			return Nan::ThrowSyntaxError(error.c_str());
+		}
+		set->sources.push_back(source);
+	}
+
+	if (!set->set.Compile())
+	{
+		return Nan::ThrowError("RE2.Set could not be compiled.");
+	}
+
+	set->combinedSource = makeCombinedSource(set->sources);
+	set->Wrap(info.This());
+	set.release();
+
+	info.GetReturnValue().Set(info.This());
+}
+
+NAN_METHOD(WrappedRE2Set::Test)
+{
+	auto re2set = Nan::ObjectWrap::Unwrap<WrappedRE2Set>(info.This());
+	if (!re2set)
+	{
+		info.GetReturnValue().Set(false);
+		return;
+	}
+
+	StrVal str;
+	v8::Local<v8::Object> keepAlive;
+	if (!fillInput(info[0], str, keepAlive))
+	{
+		return;
+	}
+
+	re2::RE2::Set::ErrorInfo errorInfo{re2::RE2::Set::kNoError};
+	bool matched = re2set->set.Match(str, nullptr, &errorInfo);
+	if (!matched && errorInfo.kind != re2::RE2::Set::kNoError)
+	{
+		const char *message = "RE2.Set matching failed.";
+		switch (errorInfo.kind)
+		{
+		case re2::RE2::Set::kOutOfMemory:
+			message = "RE2.Set matching failed: out of memory.";
+			break;
+		case re2::RE2::Set::kInconsistent:
+			message = "RE2.Set matching failed: inconsistent result.";
+			break;
+		case re2::RE2::Set::kNotCompiled:
+			message = "RE2.Set matching failed: set is not compiled.";
+			break;
+		default:
+			break;
+		}
+		return Nan::ThrowError(message);
+	}
+
+	info.GetReturnValue().Set(matched);
+}
+
+NAN_METHOD(WrappedRE2Set::Match)
+{
+	auto re2set = Nan::ObjectWrap::Unwrap<WrappedRE2Set>(info.This());
+	if (!re2set)
+	{
+		info.GetReturnValue().Set(Nan::New<v8::Array>(0));
+		return;
+	}
+
+	StrVal str;
+	v8::Local<v8::Object> keepAlive;
+	if (!fillInput(info[0], str, keepAlive))
+	{
+		return;
+	}
+
+	std::vector<int> matches;
+	re2::RE2::Set::ErrorInfo errorInfo{re2::RE2::Set::kNoError};
+	bool matched = re2set->set.Match(str, &matches, &errorInfo);
+	if (!matched && errorInfo.kind != re2::RE2::Set::kNoError)
+	{
+		const char *message = "RE2.Set matching failed.";
+		switch (errorInfo.kind)
+		{
+		case re2::RE2::Set::kOutOfMemory:
+			message = "RE2.Set matching failed: out of memory.";
+			break;
+		case re2::RE2::Set::kInconsistent:
+			message = "RE2.Set matching failed: inconsistent result.";
+			break;
+		case re2::RE2::Set::kNotCompiled:
+			message = "RE2.Set matching failed: set is not compiled.";
+			break;
+		default:
+			break;
+		}
+		return Nan::ThrowError(message);
+	}
+
+	std::sort(matches.begin(), matches.end());
+	auto result = Nan::New<v8::Array>(matches.size());
+	for (size_t i = 0, n = matches.size(); i < n; ++i)
+	{
+		Nan::Set(result, i, Nan::New(matches[i]));
```

Comment:
If I understood correctly, `matches[i]` is an index into a UTF-8 buffer. It would work for buffers but wouldn't work for strings with multi-byte (non-ASCII) characters, e.g., emojis or East-Asian symbols.

Take a look how `exec()` does it. The problem is that the process of calculating string offsets is linear, so the best bet is to sort indices and calculate distances between them forming offsets afterwards.

Obviously string offsets should be calculated only for strings. Buffers are fine.

Reply:
Yes, you are right. Indices are of matched patterns, not matches themselves.